subprojects/core/src/main/java/org/gradle/api/internal/initialization/MutableClassLoaderScope.java

```diff
@@ -0,0 +1,117 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.initialization;
+
+import org.gradle.initialization.ClassLoaderScopeId;
+import org.gradle.initialization.ClassLoaderScopeOrigin;
+import org.gradle.internal.classpath.ClassPath;
+import org.gradle.internal.hash.HashCode;
+
+import javax.annotation.Nullable;
+import java.util.function.Function;
+
+public class MutableClassLoaderScope implements ClassLoaderScope {
+
+    private ClassLoaderScope delegate;
+
+    public MutableClassLoaderScope(
+        ClassLoaderScope delegate
+    ) {
+        this.delegate = delegate;
+    }
+
+
+    @Override
+    public ClassLoaderScopeId getId() {
+        return delegate.getId();
+    }
+
+    @Nullable
+    @Override
+    public ClassLoaderScopeOrigin getOrigin() {
+        return delegate.getOrigin();
+    }
+
+    @Override
+    public ClassLoader getLocalClassLoader() {
+        return delegate.getLocalClassLoader();
+    }
+
+    @Override
+    public ClassLoader getExportClassLoader() {
+        return delegate.getExportClassLoader();
+    }
+
+    @Override
+    public ClassLoaderScope getParent() {
+        return delegate.getParent();
+    }
+
+    @Override
+    public boolean defines(Class<?> clazz) {
+        return delegate.defines(clazz);
+    }
+
+    @Override
+    public ClassLoaderScope local(ClassPath classPath) {
+        return delegate.local(classPath);
+    }
+
+    @Override
+    public ClassLoaderScope export(ClassPath classPath) {
+        return delegate.export(classPath);
+    }
+
+    @Override
+    public ClassLoaderScope export(ClassLoader classLoader) {
+        return delegate.export(classLoader);
+    }
+
+    @Override
+    public ClassLoaderScope createChild(String id, @Nullable ClassLoaderScopeOrigin origin) {
+        return delegate.createChild(id, origin);
+    }
+
+    @Override
+    public ClassLoaderScope createLockedChild(String id, @Nullable ClassLoaderScopeOrigin origin, ClassPath localClasspath, @Nullable HashCode classpathImplementationHash, @Nullable Function<ClassLoader, ClassLoader> localClassLoaderFactory) {
+        return delegate.createLockedChild(id, origin, localClasspath, classpathImplementationHash, localClassLoaderFactory);
+    }
+
+    @Override
+    public ClassLoaderScope lock() {
+        return delegate.lock();
+    }
+
+    @Override
+    public boolean isLocked() {
+        return delegate.isLocked();
+    }
+
+    @Override
+    public void onReuse() {
+        delegate.onReuse();
+    }
+
+    @Override
+    public ClassLoaderScope getOriginalScope() {
+        return delegate;
+    }
+
+    public void mutate(String childSuffix) {
```

Comment:
❌ Can we have a Javadoc on this method to describe what it does? I wonder if `mutate` is still the right work. Maybe `replaceWithMutableChild` or `asMutable` might work. WDYT?

Reply:
+1 to `asMutable()`

Reply:
Doesn't `asMutable()` sound like a method returning a mutable object?

Reply:
I agree with @snoopcheri. What about `makeMutable()`?

Reply:
We _could_ make `asMutable()` available on `ClassLoaderScope` and have it return a `MutableClassLoaderScope`.  For `DefaultClassLoaderScope` it would create/return a `MutableClassLoaderScope`.  For a `MutableClassLoaderScope`, it would return itself.  For `RootClassLoaderScope` and `ImmutableClassLoaderScope`, it would throw `UnsupportedOperationException`.

I don't feel strongly about this in any case as this isn't user-facing.  `mutate()` and `asMutable()` are my first choices.  I'm not a fan of `replaceWithMutableChild()` or `makeMutable()` but if others feel strongly, I don't object.

Reply:
The only thing I am not so sure about is what should the behavior of `DefaultClassLoaderScope` be. If it is locked, we cannot simply wrap it in a `MutableClassLoaderScope` because it would still not be mutable due to locking. Should we create a child, similar to what we do in `MutableClassLoaderScope`? For now, I think it's fine to throw an `UnsupportedOperationException` and implement it in the future if we need it.

Reply:
I was thinking `DefaultClassLoaderScope.asMutable()` would return `new MutableClassLoaderScope(this).asMutable()`.  IOW, `asMutable()` might return the same object or a new object, but whatever it returns will be mutable.  Admittedly, it's not quite right, and I'm not in love with the idea, just throwing out options.

I think leaving it unimplemented for now is fine.

Reply:
@ghale, I will merge this. If you have further feedback I will address it in a subsequent PR.